core/src/optimistically_confirmed_bank_tracker.rs

```diff
@@ -0,0 +1,278 @@
+        exit: &Arc<AtomicBool>,
+        bank_forks: Arc<RwLock<BankForks>>,
+        optimistically_confirmed_bank: Arc<RwLock<OptimisticallyConfirmedBank>>,
+        subscriptions: Arc<RpcSubscriptions>,
+    ) -> Self {
+        let exit_ = exit.clone();
+        let mut pending_optimistically_confirmed_banks = HashSet::new();
+        let thread_hdl = Builder::new()
+            .name("solana-optimistic-bank-tracker".to_string())
+            .spawn(move || loop {
+                if exit_.load(Ordering::Relaxed) {
+                    break;
+                }
+
+                if let Err(RecvTimeoutError::Disconnected) = Self::recv_notification(
+                    &receiver,
+                    &bank_forks,
+                    &optimistically_confirmed_bank,
+                    &subscriptions,
+                    &mut pending_optimistically_confirmed_banks,
+                ) {
+                    break;
+                }
+            })
+            .unwrap();
+        Self { thread_hdl }
+    }
+
+    fn recv_notification(
+        receiver: &Receiver<BankNotification>,
+        bank_forks: &Arc<RwLock<BankForks>>,
+        optimistically_confirmed_bank: &Arc<RwLock<OptimisticallyConfirmedBank>>,
+        subscriptions: &Arc<RpcSubscriptions>,
+        mut pending_optimistically_confirmed_banks: &mut HashSet<Slot>,
+    ) -> Result<(), RecvTimeoutError> {
+        let notification = receiver.recv_timeout(Duration::from_secs(1))?;
+        Self::process_notification(
+            notification,
+            bank_forks,
+            optimistically_confirmed_bank,
+            subscriptions,
+            &mut pending_optimistically_confirmed_banks,
+        );
+        Ok(())
+    }
+
+    pub(crate) fn process_notification(
+        notification: BankNotification,
+        bank_forks: &Arc<RwLock<BankForks>>,
+        optimistically_confirmed_bank: &Arc<RwLock<OptimisticallyConfirmedBank>>,
+        subscriptions: &Arc<RpcSubscriptions>,
+        pending_optimistically_confirmed_banks: &mut HashSet<Slot>,
+    ) {
+        debug!("received bank notification: {:?}", notification);
+        match notification {
+            BankNotification::OptimisticallyConfirmed(slot) => {
+                if let Some(bank) = bank_forks
+                    .read()
+                    .unwrap()
+                    .get(slot)
+                    .filter(|b| b.is_frozen())
+                {
+                    let mut w_optimistically_confirmed_bank =
+                        optimistically_confirmed_bank.write().unwrap();
+                    if bank.slot() > w_optimistically_confirmed_bank.bank.slot() {
+                        w_optimistically_confirmed_bank.bank = bank.clone();
+                    }
+                    drop(w_optimistically_confirmed_bank);
+                    subscriptions.notify_gossip_subscribers(slot);
+                } else {
+                    pending_optimistically_confirmed_banks.insert(slot);
```

Comment:
should we only add this to the pending queue if it's greater than the current optimistic bank and the current root?

Edit:  I see the `BankNotification::Root(bank)` clears outdated items from the `pending_optimistically_confirmed_banks`, but may still be good to check in case a root doesn't get set for a while